<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Asynchronous JavaScript</title>
</head>
<body>
    <h1>Asynchronous JavaScript</h1>
    <script>
        /*
        const second = () => {
            setTimeout(() => {
                console.log('Async Hey there');
            }, 2000);
        }

        const first = () => {
            console.log('Hey there!');
            second();
            console.log('The end');
        }

        first();
        */

        // CALLBACK HELL!
        // function getRecipe() {
        //     setTimeout(() => {
        //         const recipeID = [523, 685, 234, 123, 543];
        //         console.log(recipeID);

        //         setTimeout(id => {
        //             const recipe = {title: 'Fresh Tomato Pasta', publisher: 'Jonas'};
        //             console.log(`${id}: ${recipe.title}`);

        //             setTimeout(publisher => {
        //                 const recipe2 = {title: 'Italian Pizza', publisher: 'Jonas'};
        //                 console.log(recipe2);
        //             }, 1500, recipe.publisher);
        //         }, 1000, recipeID[2]);
        //     }, 1500);
        // }
        
        // function realTime() {
        //     console.log('Continuing to operate...');
        //     setTimeout(() => {
        //         console.log('And half a second later...');
        //     }, 500);
        // }

        // getRecipe();
        // realTime();

        // const getIDs = new Promise((resolve, reject) => {
        //     setTimeout(() => {
        //         resolve([523, 685, 234, 123, 543])
        //     }, 1500);
        // });

        // const getRecipe = recID => {
        //     return new Promise((resolve, reject) => {
        //         setTimeout(ID => {
        //             const recipe = {title: 'Fresh Tomato Pasta', publisher: 'Jonas'};
        //             resolve(`${ID}: ${recipe.title}`);
        //         }, 1500, recID);
        //     });
        // };

        // const getRelated = publisher => {
        //     return new Promise((resolve, reject) => {
        //         setTimeout(pub => {
        //             const recipe = {title: 'Italian Pizza', publisher: 'Jonas'};
        //             resolve(`${pub}: ${recipe.title}`);
        //         }, 1500, publisher);
        //     });
        // };

            // Since we're method-chaining, the first .then is after resolve from getIDs, the second
            // .then is from the getRecipe function call which returns a promise
        /*
        getIDs
        .then(IDs => {
            console.log(IDs);
            return getRecipe(IDs[2]);
        })
        .then(recipe => {
            console.log(recipe);
            return getRelated('Jonas');
        })
        .then(recipe => {
            console.log(recipe);
        })
        .catch(error => {
            console.log('Error!!!');
        });
        */

        // async function getRecipesAW() {
        //     console.log('Async code has started');
        //     const IDs = await getIDs;
        //     console.log(IDs);
        //     const recipe = await getRecipe(IDs[2]);
        //     console.log(recipe);
        //     const related = await getRelated('Jonas');
        //     console.log(related);
        //     console.log('Async code has ended');

        //     return recipe;
        // }
        // getRecipesAW()
        // .then(result => {
        //     console.log(`${result} is the best ever!`);
        // });

        // console.log('code has started');
        // console.log('code has ended');

            /*function getWeather(woeid) {
                fetch(`https://cors-anywhere.herokuapp.com/https://www.metaweather.com/api/location/${woeid}`)
                .then(result => {
                    //console.log(result);
                    return result.json();
                })
                .then(data => {
                    //console.log(data);
                    const today = data.consolidated_weather[0];
                    console.log(`Temperatures in ${data.title} stay between ${today.min_temp} and ${today.max_temp}`)
                })
                .catch(error => {
                    console.log(`There was an error: ${error}`);
                });
            }
            getWeather(2487956);
            getWeather(44418);
            */

            async function getWeatherAW(woeid) {
                try {
                    const result = await fetch(`https://cors-anywhere.herokuapp.com/https://www.metaweather.com/api/location/${woeid}`);
                    const data = await result.json();
                    //const today = data.consolidated_weather[0];
                    const tomorrow = data.consolidated_weather[1];
                    console.log(`Temperatures in ${data.title} stay between ${tomorrow.min_temp} and ${tomorrow.max_temp} tomorrow.`);
                    return data;
                } catch(error) {
                    console.log(error);
                }
            }
            let dataLondon; 
            getWeatherAW(2487956);
            getWeatherAW(44418).then(data => {
                dataLondon = data;
                console.log(dataLondon);
            });
    </script>
</body>
</html>